File: 0230.Kth_smallest_element_in_a_BST/solution.go
package leetcode0230

import "github.com/Ackerr/Algorithms/utils"

func kthSmallest(root *utils.TreeNode, k int) int {
	stack := make([]*utils.TreeNode, 0)
	kthVal := 0
	for root != nil || len(stack) != 0 {
		if root != nil {
			stack = append(stack, root)
			root = root.Left
		} else {
			k--
			root = stack[len(stack)-1] // python list.pop(）
			kthVal = root.Val
			stack = stack[:len(stack)-1]
			root = root.Right
		}
		if k == 0 {
			break
		}
	}
	return kthVal
}

File: 0230.Kth_smallest_element_in_a_BST/solution.py
from utils import TreeNode


class Solution:
    def kth_smallest(self, root: TreeNode, k: int) -> int:
        def middle(node):  # 二叉搜索树中序遍历得到的是递增数组
            if not node:
                return []
            return middle(node.left) + [node.val] + middle(node.right)

        return middle(root)[k - 1]
